Web Inspector: support OffscreenCanvas for Canvas related operations

dcrousso opened the pull request:
#### b5959d78adc403bde81fdad57864ec9fcf73fa21
<pre>
Web Inspector: support OffscreenCanvas for Canvas related operations
<a href="https://bugs.webkit.org/show_bug.cgi?id=180833">https://bugs.webkit.org/show_bug.cgi?id=180833</a>
&lt;rdar://problem/36059660&gt;

Reviewed by Patrick Angle.

* Source/JavaScriptCore/inspector/protocol/Canvas.json:
* Source/JavaScriptCore/inspector/protocol/Recording.json:
Expose the entire domain (minus a few commands/events) to the `&quot;worker&quot;` target.

* Source/WebCore/inspector/InspectorController.cpp:
(WebCore::InspectorController::createLazyAgents):
* Source/WebCore/inspector/WorkerInspectorController.cpp:
(WebCore::WorkerInspectorController::createLazyAgents):
* Source/WebCore/inspector/agents/InspectorCanvasAgent.h:
* Source/WebCore/inspector/agents/InspectorCanvasAgent.cpp:
(WebCore::InspectorCanvasAgent::InspectorCanvasAgent):
(WebCore::InspectorCanvasAgent::enable):
(WebCore::InspectorCanvasAgent::disable):
(WebCore::InspectorCanvasAgent::enabled const): Added.
(WebCore::InspectorCanvasAgent::internalEnable): Added.
(WebCore::InspectorCanvasAgent::internalDisable): Added.
(WebCore::InspectorCanvasAgent::didChangeCanvasSize): Added.
(WebCore::InspectorCanvasAgent::didChangeCanvasMemory):
(WebCore::InspectorCanvasAgent::unbindCanvas):
(WebCore::InspectorCanvasAgent::requestNode): Deleted.
(WebCore::InspectorCanvasAgent::requestClientNodes): Deleted.
(WebCore::InspectorCanvasAgent::frameNavigated): Deleted.
(WebCore::InspectorCanvasAgent::didChangeCSSCanvasClientNodes): Deleted.
* Source/WebCore/inspector/agents/page/PageCanvasAgent.h: Added.
* Source/WebCore/inspector/agents/page/PageCanvasAgent.cpp: Added.
(WebCore::PageCanvasAgent::PageCanvasAgent):
(WebCore::PageCanvasAgent::enabled const):
(WebCore::PageCanvasAgent::internalEnable):
(WebCore::PageCanvasAgent::internalDisable):
(WebCore::PageCanvasAgent::requestNode):
(WebCore::PageCanvasAgent::requestClientNodes):
(WebCore::PageCanvasAgent::frameNavigated):
(WebCore::PageCanvasAgent::didChangeCSSCanvasClientNodes):
(WebCore::PageCanvasAgent::matchesCurrentContext const):
* Source/WebCore/inspector/agents/worker/WorkerCanvasAgent.h: Added.
* Source/WebCore/inspector/agents/worker/WorkerCanvasAgent.cpp: Added.
(WebCore::WorkerCanvasAgent::WorkerCanvasAgent):
(WebCore::WorkerCanvasAgent::requestNode):
(WebCore::WorkerCanvasAgent::requestClientNodes):
(WebCore::WorkerCanvasAgent::matchesCurrentContext const):
Split `InspectorCanvasAgent` into `PageCanvasAgent` and `WorkerCanvasAgent`.

* Source/WebInspectorUI/UserInterface/Protocol/CanvasObserver.js:
(WI.CanvasObserver.prototype.canvasAdded):
(WI.CanvasObserver.prototype.canvasRemoved):
(WI.CanvasObserver.prototype.canvasSizeChanged):
(WI.CanvasObserver.prototype.canvasMemoryChanged):
(WI.CanvasObserver.prototype.clientNodesChanged):
(WI.CanvasObserver.prototype.recordingStarted):
(WI.CanvasObserver.prototype.recordingProgress):
(WI.CanvasObserver.prototype.recordingFinished):
(WI.CanvasObserver.prototype.extensionEnabled):
(WI.CanvasObserver.prototype.programCreated):
(WI.CanvasObserver.prototype.programDeleted):
(WI.CanvasObserver.prototype.cssCanvasClientNodesChanged):
Pass along the `WI.Target`.

* Source/WebInspectorUI/UserInterface/Controllers/CanvasManager.js:
(WI.CanvasManager):
(WI.CanvasManager.prototype.disable):
(WI.CanvasManager.prototype.canvasAdded):
(WI.CanvasManager.prototype.canvasRemoved):
(WI.CanvasManager.prototype.canvasSizeChanged): Added.
(WI.CanvasManager.prototype.canvasMemoryChanged):
(WI.CanvasManager.prototype.clientNodesChanged):
(WI.CanvasManager.prototype.recordingStarted):
(WI.CanvasManager.prototype.recordingProgress):
(WI.CanvasManager.prototype.recordingFinished):
(WI.CanvasManager.prototype.extensionEnabled):
(WI.CanvasManager.prototype.programCreated):
(WI.CanvasManager.prototype.programDeleted):
(WI.CanvasManager.prototype._canvasForIdentifier): Added.
(WI.CanvasManager.prototype._handleTargetRemoved): Added.
(WI.CanvasManager.prototype._mainResourceDidChange):
(WI.CanvasManager.prototype.get shaderPrograms):
`WI.Canvas`/`WI.ShaderProgram` now need to be keyed by a `WI.Target` -&gt; identifier combo since the same identifier could be used by two different `WI.Target` (e.g. main page and `Worker`).

* Source/WebInspectorUI/UserInterface/Models/Canvas.js:
(WI.Canvas):
(WI.Canvas.fromPayload):
(WI.Canvas.prototype.get target): Added.
(WI.Canvas.prototype.get size): Added.
(WI.Canvas.prototype.get memoryCost):
(WI.Canvas.prototype.requestNode):
(WI.Canvas.prototype.requestContent):
(WI.Canvas.prototype.requestClientNodes):
(WI.Canvas.prototype.startRecording):
(WI.Canvas.prototype.stopRecording):
(WI.Canvas.prototype.sizeChanged): Added.
(WI.Canvas.prototype.memoryChanged): Renamed from `set memoryCost`.
(WI.Canvas.prototype.async _calculateSize): Added.
(WI.Canvas.prototype.requestSize): Deleted.
* Source/WebInspectorUI/UserInterface/Models/ShaderProgram.js:
(WI.ShaderProgram):
(WI.ShaderProgram.prototype.get target): Added.
(WI.ShaderProgram.prototype.set disabled):
(WI.ShaderProgram.prototype.requestShaderSource):
(WI.ShaderProgram.prototype.updateShader):
(WI.ShaderProgram.prototype.showHighlight):
(WI.ShaderProgram.prototype.hideHighlight):
Use the `this._target` instead of `WI.assumingMainTarget()`.

* Source/WebCore/html/CanvasBase.h:
(WebCore::CanvasBase::setSize): Deleted.
* Source/WebCore/html/CanvasBase.cpp:
(WebCore::CanvasBase::setSize): Added.
(WebCore::CanvasBase::setImageBuffer const):
* Source/WebCore/inspector/InspectorCanvas.cpp:
(WebCore::InspectorCanvas::buildObjectForCanvas):
* Source/WebCore/inspector/InspectorInstrumentation.h:
(WebCore::InspectorInstrumentation::didChangeCanvasSize): Added.
* Source/WebCore/inspector/InspectorInstrumentation.cpp:
(WebCore::InspectorInstrumentation::didCommitLoadImpl):
(WebCore::InspectorInstrumentation::didChangeCSSCanvasClientNodesImpl):
(WebCore::InspectorInstrumentation::didChangeCanvasSizeImpl): Added.
* Source/WebCore/inspector/InstrumentingAgents.h:
* Source/WebInspectorUI/UserInterface/Views/CanvasContentView.js:
(WI.CanvasContentView):
(WI.CanvasContentView.prototype.initialLayout):
(WI.CanvasContentView.prototype.attached):
(WI.CanvasContentView.prototype.detached):
(WI.CanvasContentView.prototype._populateCanvasElementButtonContextMenu):
(WI.CanvasContentView.prototype._updateSize): Added.
(WI.CanvasContentView.prototype._refreshPixelSize): Deleted.
Now that `Canvas.requestNode` won&apos;t always work, we need a different way of getting the owner to find it&apos;s size.
Additionally, since the owner might not be a `Element`, we can&apos;t leverage `WI.DOMNode.Event.AttributeModified` to listen for changes in the size.
Instead, add instrumentation for when the size changes in C++ and notify the frontend as such (as well as adding logic to grab the `context.canvas.{width,height}` from the `Canvas.resolveCanvasContext` instead since that will always work).

* Source/WebInspectorUI/UserInterface/Models/Geometry.js:
(WI.Size.fromJSON):
Drive-by: Add a helper to make some logic slightly easier.

* Source/WebInspectorUI/UserInterface/Models/Recording.js:
(WI.Recording.displayNameForRecordingType):
(WI.Recording.prototype.is2D):
Drive-by: Fix display name to be consistent with `WI.Canvas.displayNameForContextType`.
Drive-by: Add missing method.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:

* LayoutTests/inspector/canvas/resources/create-context-utilities.js:
(destroyCanvases):
(InspectorTest.CreateContextUtilities.addSimpleTestCase):
* LayoutTests/inspector/canvas/resources/worker.js: Added.
(createContext):
(destroyContexts):
* LayoutTests/inspector/canvas/create-context-2d.html:
* LayoutTests/inspector/canvas/create-context-2d-expected.txt:
* LayoutTests/inspector/canvas/create-context-bitmaprenderer.html:
* LayoutTests/inspector/canvas/create-context-bitmaprenderer-expected.txt:
* LayoutTests/inspector/canvas/create-context-webgl.html:
* LayoutTests/inspector/canvas/create-context-webgl-expected.txt:
* LayoutTests/inspector/canvas/create-context-webgl2.html:
* LayoutTests/inspector/canvas/create-context-webgl2-expected.txt:

Canonical link: <a href="https://commits.webkit.org/267488@main">https://commits.webkit.org/267488@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/91dde642d4bcffb9677b61f833634afa2a365c64

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/16741 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/17064 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/17515 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/18522 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/15682 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/16930 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/20296 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/17203 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/17971 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/16940 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/17306 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/14487 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/19311 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/14550 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/15171 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/21924 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/14439 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/15539 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/15337 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/19639 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/15970 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/15942 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/13522 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/18299 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/15119 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/15011 "Passed tests") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/4280 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4008 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/19484 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/19522 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/15772 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/4127 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

rcaliman-apple commented on Source/WebInspectorUI/UserInterface/Controllers/CanvasManager.js:
```diff
@@ -176,82 +178,132 @@ WI.CanvasManager = class CanvasManager extends WI.Object
+        if (!canvasForIdentifierMap)
+            return;
+
+        let canvas = canvasForIdentifierMap.get(canvasIdentifier);
```
This block seems repeated frequently in `WI.CanvasManager`. Do you think it makes sense to extract it to a method like `getCanvasForCanvasIdentifierAndTarget(canvasIdentifier, target)` or similarly named? 

It would help keep all the other methods leaner and easier to read.

rcaliman-apple commented on Source/WebInspectorUI/UserInterface/Controllers/CanvasManager.js:
```diff
@@ -156,18 +151,25 @@ WI.CanvasManager = class CanvasManager extends WI.Object
-        console.assert(!this._canvasIdentifierMap.has(canvasPayload.canvasId), `Canvas already exists with id ${canvasPayload.canvasId}.`);
+        let canvas = WI.Canvas.fromPayload(target, canvasPayload);
+
+        let canvasForIdentifierMap = this._canvasForIdentifierForTargetMap.getOrInitialize(target, () => new Map);
```
What's the reason for initializing with the `new Map` returned by the function instead of just passing `new Map`? 
Is it so the new instance is created only when needed instead of every time the `Map.getOrInitialize()` is called?

patrickangle replied:
Your assumption is correct, by wrapping in a function we avoid creating the map if it already exists. That said this is a balance between creating a function or creating a map, neither of which will be used if it already exists... I'd be curious to know which creation is "heavier" memory wise 😅

rcaliman-apple commented on Source/WebInspectorUI/UserInterface/Models/Canvas.js:
```diff
@@ -53,11 +57,24 @@ WI.Canvas = class Canvas extends WI.Object
         this._recordingFrames = [];
         this._recordingBufferUsed = 0;
+
+        // COMPATIBILITY (macOS X.0, iOS X.0): `Canvas.canvasSizeChanged` did not exist yet.
```
The compatibility comment should likely become 
```
// COMPATIBILITY (macOS 14.0, iOS 17.0)
```
Ditto everywhere else.

dcrousso replied:
i dont see a protocol file for macOS 14 or iOS 17, so it's still technically correct as-is

rcaliman-apple commented on Source/WebInspectorUI/UserInterface/Protocol/CanvasObserver.js:
```diff
@@ -29,42 +29,47 @@ WI.CanvasObserver = class CanvasObserver extends InspectorBackend.Dispatcher
+        WI.canvasManager.canvasRemoved(this._target, canvasId);
+    }
+
+    canvasSizeChanged(canvasId, width, height)
```
Should we have a compatibility comment here?

```
// COMPATIBILITY (macOS 14.0, iOS 17.0) `Canvas.canvasSizeChanged` did not exist yet.
```

dcrousso replied:
we dont normally add compatibility comments for events since they're only called from the backend (unless we're using it as a check for something)

rcaliman-apple commented on Source/WebInspectorUI/UserInterface/Models/Canvas.js:
```diff
@@ -217,8 +226,12 @@ WI.Canvas = class Canvas extends WI.Object
 
-                let target = WI.assumingMainTarget();
-                target.CanvasAgent.requestNode(this._identifier, (error, nodeId) => {
+                if (!this._target.hasCommand("Canvas.requestNode")) {
+                    resolve(null);
+                    return;
+                }
```
Thanks for checking the target capabilities! 👍

rcaliman-apple commented on Source/WebInspectorUI/UserInterface/Models/Recording.js:
```diff
@@ -335,6 +335,11 @@ WI.Recording = class Recording extends WI.Object
         this._displayName = name;
     }
 
+    is2D()
+    {
+        return WI.Recording.is2D(this._type);
+    }
+
```
This is clever! I hadn't realized that a `static` method doesn't get overwritten by an identically named method on the object instance. 🤯

rcaliman-apple reviewed (approved):
This patch is associated with https://bugs.webkit.org/show_bug.cgi?id=180833 which is already marked as RESOLVED FIXED by the identically named PR https://github.com/WebKit/WebKit/pull/11648. Did you mean to create another Bugzilla issue for it,  perhaps mentioning the focus on workers?

The changes related to Web Inspector, both frontend and backend, look good. 👍 

Can we have @kkinnunen-apple review the changes in `WebCore`  that pertain to Canvas and Worker modifications? 🙏 

Pinging @xeenon or @patrickangle for review/approval.

patrickangle reviewed (approved):
Are there more canvas tests that we could augment to run in both the main and a worker context so that as tests are added folks assume they should be testing both?

dcrousso reviewed (commented):
> This patch is associated with https://bugs.webkit.org/show_bug.cgi?id=180833 which is already marked as RESOLVED FIXED by the identically named PR #11648. Did you mean to create another Bugzilla issue for it, perhaps mentioning the focus on workers?

#11648 took parts of this PR when it was still a draft, but not all of it.  i didnt get a chance to finish this (draft) PR until after it landed

> Are there more canvas tests that we could augment to run in both the main and a worker context so that as tests are added folks assume they should be testing both?

rather than create separate tests for workers, i felt it was easier to just include the worker bits in the existing tests.  but this PR only makes the current `OffscreenCanvas` support work inside `Worker`.  it doesnt expand the support `OffscreenCanvas` (e.g. other context types, recording, etc.).  that's all done by #13845

dcrousso committed:
Web Inspector: support OffscreenCanvas for Canvas related operations
https://bugs.webkit.org/show_bug.cgi?id=180833
<rdar://problem/36059660>

Reviewed by Patrick Angle.

* Source/JavaScriptCore/inspector/protocol/Canvas.json:
* Source/JavaScriptCore/inspector/protocol/Recording.json:
Expose the entire domain (minus a few commands/events) to the `"worker"` target.

* Source/WebCore/inspector/InspectorController.cpp:
(WebCore::InspectorController::createLazyAgents):
* Source/WebCore/inspector/WorkerInspectorController.cpp:
(WebCore::WorkerInspectorController::createLazyAgents):
* Source/WebCore/inspector/agents/InspectorCanvasAgent.h:
* Source/WebCore/inspector/agents/InspectorCanvasAgent.cpp:
(WebCore::InspectorCanvasAgent::InspectorCanvasAgent):
(WebCore::InspectorCanvasAgent::enable):
(WebCore::InspectorCanvasAgent::disable):
(WebCore::InspectorCanvasAgent::enabled const): Added.
(WebCore::InspectorCanvasAgent::internalEnable): Added.
(WebCore::InspectorCanvasAgent::internalDisable): Added.
(WebCore::InspectorCanvasAgent::didChangeCanvasSize): Added.
(WebCore::InspectorCanvasAgent::didChangeCanvasMemory):
(WebCore::InspectorCanvasAgent::unbindCanvas):
(WebCore::InspectorCanvasAgent::requestNode): Deleted.
(WebCore::InspectorCanvasAgent::requestClientNodes): Deleted.
(WebCore::InspectorCanvasAgent::frameNavigated): Deleted.
(WebCore::InspectorCanvasAgent::didChangeCSSCanvasClientNodes): Deleted.
* Source/WebCore/inspector/agents/page/PageCanvasAgent.h: Added.
* Source/WebCore/inspector/agents/page/PageCanvasAgent.cpp: Added.
(WebCore::PageCanvasAgent::PageCanvasAgent):
(WebCore::PageCanvasAgent::enabled const):
(WebCore::PageCanvasAgent::internalEnable):
(WebCore::PageCanvasAgent::internalDisable):
(WebCore::PageCanvasAgent::requestNode):
(WebCore::PageCanvasAgent::requestClientNodes):
(WebCore::PageCanvasAgent::frameNavigated):
(WebCore::PageCanvasAgent::didChangeCSSCanvasClientNodes):
(WebCore::PageCanvasAgent::matchesCurrentContext const):
* Source/WebCore/inspector/agents/worker/WorkerCanvasAgent.h: Added.
* Source/WebCore/inspector/agents/worker/WorkerCanvasAgent.cpp: Added.
(WebCore::WorkerCanvasAgent::WorkerCanvasAgent):
(WebCore::WorkerCanvasAgent::requestNode):
(WebCore::WorkerCanvasAgent::requestClientNodes):
(WebCore::WorkerCanvasAgent::matchesCurrentContext const):
Split `InspectorCanvasAgent` into `PageCanvasAgent` and `WorkerCanvasAgent`.

* Source/WebInspectorUI/UserInterface/Protocol/CanvasObserver.js:
(WI.CanvasObserver.prototype.canvasAdded):
(WI.CanvasObserver.prototype.canvasRemoved):
(WI.CanvasObserver.prototype.canvasSizeChanged):
(WI.CanvasObserver.prototype.canvasMemoryChanged):
(WI.CanvasObserver.prototype.clientNodesChanged):
(WI.CanvasObserver.prototype.recordingStarted):
(WI.CanvasObserver.prototype.recordingProgress):
(WI.CanvasObserver.prototype.recordingFinished):
(WI.CanvasObserver.prototype.extensionEnabled):
(WI.CanvasObserver.prototype.programCreated):
(WI.CanvasObserver.prototype.programDeleted):
(WI.CanvasObserver.prototype.cssCanvasClientNodesChanged):
Pass along the `WI.Target`.

* Source/WebInspectorUI/UserInterface/Controllers/CanvasManager.js:
(WI.CanvasManager):
(WI.CanvasManager.prototype.disable):
(WI.CanvasManager.prototype.canvasAdded):
(WI.CanvasManager.prototype.canvasRemoved):
(WI.CanvasManager.prototype.canvasSizeChanged): Added.
(WI.CanvasManager.prototype.canvasMemoryChanged):
(WI.CanvasManager.prototype.clientNodesChanged):
(WI.CanvasManager.prototype.recordingStarted):
(WI.CanvasManager.prototype.recordingProgress):
(WI.CanvasManager.prototype.recordingFinished):
(WI.CanvasManager.prototype.extensionEnabled):
(WI.CanvasManager.prototype.programCreated):
(WI.CanvasManager.prototype.programDeleted):
(WI.CanvasManager.prototype._canvasForIdentifier): Added.
(WI.CanvasManager.prototype._handleTargetRemoved): Added.
(WI.CanvasManager.prototype._mainResourceDidChange):
(WI.CanvasManager.prototype.get shaderPrograms):
`WI.Canvas`/`WI.ShaderProgram` now need to be keyed by a `WI.Target` -> identifier combo since the same identifier could be used by two different `WI.Target` (e.g. main page and `Worker`).

* Source/WebInspectorUI/UserInterface/Models/Canvas.js:
(WI.Canvas):
(WI.Canvas.fromPayload):
(WI.Canvas.prototype.get target): Added.
(WI.Canvas.prototype.get size): Added.
(WI.Canvas.prototype.get memoryCost):
(WI.Canvas.prototype.requestNode):
(WI.Canvas.prototype.requestContent):
(WI.Canvas.prototype.requestClientNodes):
(WI.Canvas.prototype.startRecording):
(WI.Canvas.prototype.stopRecording):
(WI.Canvas.prototype.sizeChanged): Added.
(WI.Canvas.prototype.memoryChanged): Renamed from `set memoryCost`.
(WI.Canvas.prototype.async _calculateSize): Added.
(WI.Canvas.prototype.requestSize): Deleted.
* Source/WebInspectorUI/UserInterface/Models/ShaderProgram.js:
(WI.ShaderProgram):
(WI.ShaderProgram.prototype.get target): Added.
(WI.ShaderProgram.prototype.set disabled):
(WI.ShaderProgram.prototype.requestShaderSource):
(WI.ShaderProgram.prototype.updateShader):
(WI.ShaderProgram.prototype.showHighlight):
(WI.ShaderProgram.prototype.hideHighlight):
Use the `this._target` instead of `WI.assumingMainTarget()`.

* Source/WebCore/html/CanvasBase.h:
(WebCore::CanvasBase::setSize): Deleted.
* Source/WebCore/html/CanvasBase.cpp:
(WebCore::CanvasBase::setSize): Added.
(WebCore::CanvasBase::setImageBuffer const):
* Source/WebCore/inspector/InspectorCanvas.cpp:
(WebCore::InspectorCanvas::buildObjectForCanvas):
* Source/WebCore/inspector/InspectorInstrumentation.h:
(WebCore::InspectorInstrumentation::didChangeCanvasSize): Added.
* Source/WebCore/inspector/InspectorInstrumentation.cpp:
(WebCore::InspectorInstrumentation::didCommitLoadImpl):
(WebCore::InspectorInstrumentation::didChangeCSSCanvasClientNodesImpl):
(WebCore::InspectorInstrumentation::didChangeCanvasSizeImpl): Added.
* Source/WebCore/inspector/InstrumentingAgents.h:
* Source/WebInspectorUI/UserInterface/Views/CanvasContentView.js:
(WI.CanvasContentView):
(WI.CanvasContentView.prototype.initialLayout):
(WI.CanvasContentView.prototype.attached):
(WI.CanvasContentView.prototype.detached):
(WI.CanvasContentView.prototype._populateCanvasElementButtonContextMenu):
(WI.CanvasContentView.prototype._updateSize): Added.
(WI.CanvasContentView.prototype._refreshPixelSize): Deleted.
Now that `Canvas.requestNode` won't always work, we need a different way of getting the owner to find it's size.
Additionally, since the owner might not be a `Element`, we can't leverage `WI.DOMNode.Event.AttributeModified` to listen for changes in the size.
Instead, add instrumentation for when the size changes in C++ and notify the frontend as such (as well as adding logic to grab the `context.canvas.{width,height}` from the `Canvas.resolveCanvasContext` instead since that will always work).

* Source/WebInspectorUI/UserInterface/Models/Geometry.js:
(WI.Size.fromJSON):
Drive-by: Add a helper to make some logic slightly easier.

* Source/WebInspectorUI/UserInterface/Models/Recording.js:
(WI.Recording.displayNameForRecordingType):
(WI.Recording.prototype.is2D):
Drive-by: Fix display name to be consistent with `WI.Canvas.displayNameForContextType`.
Drive-by: Add missing method.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:

* LayoutTests/inspector/canvas/resources/create-context-utilities.js:
(destroyCanvases):
(InspectorTest.CreateContextUtilities.addSimpleTestCase):
* LayoutTests/inspector/canvas/resources/worker.js: Added.
(createContext):
(destroyContexts):
* LayoutTests/inspector/canvas/create-context-2d.html:
* LayoutTests/inspector/canvas/create-context-2d-expected.txt:
* LayoutTests/inspector/canvas/create-context-bitmaprenderer.html:
* LayoutTests/inspector/canvas/create-context-bitmaprenderer-expected.txt:
* LayoutTests/inspector/canvas/create-context-webgl.html:
* LayoutTests/inspector/canvas/create-context-webgl-expected.txt:
* LayoutTests/inspector/canvas/create-context-webgl2.html:
* LayoutTests/inspector/canvas/create-context-webgl2-expected.txt:

Canonical link: https://commits.webkit.org/267488@main
